File: app/src/main/java/com/kkkkkn/readbooks/util/ChapterUtil.java
package com.kkkkkn.readbooks.util;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;

//章节内容缓存
public class ChapterUtil {
    private final static String TAG="ChapterUtil";
    //缓存根目录
    private static final String filePath="/cacheChapters/";

    //保存读取到的缓存文字 章节URL ，后期可加密
    public static boolean cacheChapter(JSONArray jsonArray,final String path, final String bookName,final  String chapterName){
        if(jsonArray==null||bookName.isEmpty()||chapterName.isEmpty()){
            Log.e(TAG, "cacheChapter: 入参为空");
            return false;
        }
        FileOutputStream out = null;
        BufferedWriter writer = null;
        try{
            File file=new File(path+filePath+"/"+bookName+"/");
            if(!file.exists()){
                file.mkdirs();
            }
            out = new FileOutputStream(new File(path+filePath+"/"+bookName+"/"+chapterName));
            writer = new BufferedWriter(new OutputStreamWriter(out));
            for (int i = 0; i < jsonArray.length(); i++) {
                writer.write(jsonArray.getString(i));
                writer.newLine();
            }
            writer.flush();
            return true;
        }catch (IOException | JSONException e){
            e.printStackTrace();
            return false;
        }finally {
            try{
                if(writer != null){
                    writer.close();
                }
            }catch (IOException e){
                e.printStackTrace();
            }
        }
    }

    //读取缓存的章节文字
    public static String[] readCacheChapter(final String path,final String fileName,final String chapterName){
        if(fileName.isEmpty()){
            Log.e(TAG, "cacheChapter: 入参为空");
            return null;
        }
        FileInputStream inputStream = null;
        BufferedReader reader = null;
        ArrayList<String> arrayList=new ArrayList<String>();
        try{
            File file=new File(path+filePath+fileName+"/"+chapterName);
            if(!file.exists()||!file.isFile()){
                Log.e(TAG, "readCacheChapter: 未找到缓存文件");
                return null;
            }
            inputStream = new FileInputStream(file);
            reader = new BufferedReader(new InputStreamReader(inputStream));
            String temp="";
            while ((temp=reader.readLine())!=null){
                arrayList.add(temp);
            }
            int len=arrayList.size();
            return (String[])arrayList.toArray(new String[len]);
        }catch (IOException e){
            e.printStackTrace();
            return null;
        }finally {
            try{
                if(reader != null){
                    reader.close();
                }
            }catch (IOException e){
                e.printStackTrace();
            }
        }
    }

}

File: app/src/main/java/com/kkkkkn/readbooks/util/ChapterUtil.kt
package com.kkkkkn.readbooks.util

import android.util.Log
import org.json.JSONArray
import org.json.JSONException
import java.io.*

//章节内容缓存
object ChapterUtil {
    private const val TAG = "ChapterUtil"

    //缓存根目录
    private const val filePath = "/cacheChapters/"

    //保存读取到的缓存文字 章节URL ，后期可加密
    fun cacheChapter(
        jsonArray: JSONArray?,
        path: String,
        bookName: String,
        chapterName: String
    ): Boolean {
        if (jsonArray == null || bookName.isEmpty() || chapterName.isEmpty()) {
            Log.e(TAG, "cacheChapter: 入参为空")
            return false
        }
        val out: FileOutputStream?
        var writer: BufferedWriter? = null
        return try {
            val file = File("$path$filePath/$bookName/")
            if (!file.exists()) {
                file.mkdirs()
            }
            out = FileOutputStream(File("$path$filePath/$bookName/$chapterName"))
            writer = BufferedWriter(OutputStreamWriter(out))
            for (i in 0 until jsonArray.length()) {
                writer.write(jsonArray.getString(i))
                writer.newLine()
            }
            writer.flush()
            true
        } catch (e: IOException) {
            e.printStackTrace()
            false
        } catch (e: JSONException) {
            e.printStackTrace()
            false
        } finally {
            try {
                writer?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    //读取缓存的章节文字
    fun readCacheChapter(path: String, fileName: String?, chapterName: String?): Array<String>? {
        if (fileName==null||fileName.isEmpty()||chapterName==null||chapterName.isEmpty()) {
            Log.e(TAG, "cacheChapter: 入参为空")
            return null
        }
        val inputStream: FileInputStream?
        var reader: BufferedReader? = null
        val arrayList = ArrayList<String>()
        return try {
            val file = File("$path$filePath$fileName/$chapterName")
            if (!file.exists() || !file.isFile) {
                Log.e(TAG, "readCacheChapter: 未找到缓存文件")
                return null
            }
            Log.i(TAG, "readCacheChapter: "+file.path)
            inputStream = FileInputStream(file)
            reader = BufferedReader(InputStreamReader(inputStream))
            var temp: String?
            while (reader.readLine().also { temp = it } != null) {
                temp?.let { arrayList.add(it) }
            }
            arrayList.toTypedArray()
        } catch (e: IOException) {
            e.printStackTrace()
            null
        } finally {
            try {
                reader?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }
}